btree/btree.c

```diff
@@ -0,0 +1,13037 @@
+}
+
+/**
+ * Returns the CPU addressable pointer from the formatted segment address.
+ *
+ * @param seg_addr points to the formatted segment address.
+ *
+ * @return CPU addressable value.
+ */
+static void* segaddr_addr(const struct segaddr *seg_addr)
+{
+	return (void *)(seg_addr->as_core);
+}
+
+/**
+ * Returns the node type stored at segment address.
+ *
+ * @param seg_addr points to the formatted segment address.
+ *
+ * @return Node type.
+ */
+uint32_t segaddr_ntype_get(const struct segaddr *addr)
+{
+	struct node_header *h  =  segaddr_addr(addr) +
+				  sizeof(struct m0_format_header);
+
+	M0_IN(h->h_node_type, (BNT_FIXED_FORMAT,
+			       BNT_FIXED_KEYSIZE_VARIABLE_VALUESIZE,
+			       BNT_VARIABLE_KEYSIZE_VARIABLE_VALUESIZE));
+	return h->h_node_type;
+}
+
+static const struct node_type fixed_format;
+static const struct node_type fixed_ksize_variable_vsize_format;
+static const struct node_type variable_kv_format;
+
+static const struct node_type *btree_node_format[] = {
+	[BNT_FIXED_FORMAT]                        = &fixed_format,
+	[BNT_FIXED_KEYSIZE_VARIABLE_VALUESIZE]    = &fixed_ksize_variable_vsize_format,
+	[BNT_VARIABLE_KEYSIZE_FIXED_VALUESIZE]    = NULL,
+	[BNT_VARIABLE_KEYSIZE_VARIABLE_VALUESIZE] = &variable_kv_format,
+};
+
+
+/**
+ * Locates a tree descriptor whose root node points to the node at addr and
+ * return this tree to the caller.
+ * If an existing tree descriptor pointing to this root node is not found then
+ * a new tree descriptor is allocated from the free pool and the root node is
+ * assigned to this new tree descriptor.
+ * If root node pointer is not provided then this function will just allocate a
+ * tree descriptor and return it to the caller. This functionality currently is
+ * used by the create_tree function.
+ *
+ * @param op is used to exchange operation parameters and return values..
+ * @param addr is the segment address of root node.
+ * @param nxt is the next state to be returned to the caller.
+ *
+ * @return Next state to proceed in.
+ */
+static int64_t tree_get(struct node_op *op, struct segaddr *addr, int nxt)
+{
+	struct td              *tree = NULL;
+	struct nd              *node = NULL;
+
+	if (addr != NULL) {
+		nxt  = bnode_get(op, NULL, addr, nxt);
+		if (op->no_op.o_sm.sm_rc < 0)
+			return op->no_op.o_sm.sm_rc;
+		node = op->no_node;
+		m0_rwlock_write_lock(&list_lock);
+		tree = node->n_tree;
+
+		if (tree == NULL) {
+			tree = m0_alloc(sizeof *tree);
+			m0_rwlock_init(&tree->t_lock);
+			m0_rwlock_write_lock(&tree->t_lock);
+
+			tree->t_ref = 1;
+			tree->t_root = node;
+			tree->t_height = bnode_level(node) + 1;
+			bnode_fid(node, &tree->t_fid);
+			bnode_lock(node);
+			node->n_tree = tree;
+			bnode_unlock(node);
+		} else {
+			m0_rwlock_write_lock(&tree->t_lock);
+			tree->t_ref++;
+		}
+
+		op->no_node = tree->t_root;
+		op->no_tree = tree;
+		m0_rwlock_write_unlock(&tree->t_lock);
+		m0_rwlock_write_unlock(&list_lock);
+	} else
+		return M0_ERR(-EINVAL);
```

Comment:
Caller expecting next state to be returned, will returning -EINVAL cause any issue?

Reply:
Corrected.